File: dotnet/CardsProject/Domain/Content/DraftSession.cs
using CardsProject.Domain.Cards;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace CardsProject.Domain.Content;

public enum DraftSessionStatusType
{
    WaitingForPlayers,
    Drafting,
    Completed,
    Abandoned
}

public enum DraftSessionDraftTypeType
{
    Booster,
    Cube,
    Rochester
}

public class DraftSession : IValidatableObject
{
    public int Id { get; set; }

    public DraftSessionStatusType Status { get; set; }
    public DraftSessionDraftTypeType DraftType { get; set; }
    public int Seats { get; set; } = 8;
    public int TimePerPickSeconds { get; set; } = 30;
    [JsonPropertyName("createdAt")]
    public DateTime? CreatedAt { get; set; } = null;
    [JsonPropertyName("completedAt")]
    public DateTime? CompletedAt { get; set; } = null;

    public int? CardSetId { get; set; }
    [ForeignKey(nameof(CardSetId))]
    public CardSet? CardSet { get; set; }

    // Business operations

    public void Start()
    {
        // TODO: implement start
    }

    public void Abandon()
    {
        // TODO: implement abandon
    }

    public void Complete()
    {
        // TODO: implement complete
    }

    public bool IsFull()
    {
        // TODO: implement is_full
        return default;
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static readonly System.Collections.Generic.Dictionary<DraftSessionStatusType, DraftSessionStatusType[]> AllowedTransitions = new()
    {
        [DraftSessionStatusType.WaitingForPlayers] = new[] { DraftSessionStatusType.Drafting, DraftSessionStatusType.Abandoned },
        [DraftSessionStatusType.Drafting] = new[] { DraftSessionStatusType.Completed, DraftSessionStatusType.Abandoned }
    };

    public void AssertTransition(DraftSessionStatusType to)
    {
        if (!AllowedTransitions.TryGetValue(Status, out var allowed) || !System.Array.Exists(allowed, s => s == to))
            throw new InvalidOperationException($"Transition {Status} -> {to} not allowed");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( Seats >= 2 && Seats <= 16 ))
            yield return new ValidationResult("Draft session must have between 2 and 16 seats", new[] { nameof(Id) });
        if (!( TimePerPickSeconds > 0 ))
            yield return new ValidationResult("Time per pick must be greater than zero", new[] { nameof(Id) });
    }
}

File: phoenix/lib/cards_project/content/draft_session.ex
defmodule CardsProject.Content.DraftSession do
  use Ecto.Schema
  import Ecto.Changeset

  schema "draft_sessions" do
    field :status, :string
    field :draft_type, :string
    field :seats, :integer, default: 8
    field :time_per_pick_seconds, :integer, default: 30
    field :created_at, :naive_datetime
    field :completed_at, :naive_datetime
    belongs_to :card_set, CardsProject.Cards.CardSet
    has_many :participants, CardsProject.Content.DraftParticipant, foreign_key: :session_id

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:seats, :time_per_pick_seconds, :created_at, :status, :draft_type, :completed_at, :card_set_id])
    |> validate_required([:seats, :time_per_pick_seconds, :created_at])
    |> validate_inclusion(:status, ["WaitingForPlayers", "Drafting", "Completed", "Abandoned"])
    |> validate_inclusion(:draft_type, ["Booster", "Cube", "Rochester"])
    |> validate_number(:seats, greater_than_or_equal_to: 2, less_than_or_equal_to: 16, message: "Draft session must have between 2 and 16 seats")
    |> validate_number(:time_per_pick_seconds, greater_than: 0, message: "Time per pick must be greater than zero")
    |> then(fn cs ->
      if not is_nil(get_field(cs, :completed_at)) and (get_field(cs, :status) != "Completed") do
        Ecto.Changeset.add_error(cs, :status, "completed_at can only be set when draft status is Completed")
      else
        cs
      end
    end)
  end

  # ── Business operations ────────────────────────────────────────────

  def start(_record) do
    # TODO: implement DraftSession.start
    :ok
  end

  def abandon(_record) do
    # TODO: implement DraftSession.abandon
    :ok
  end

  def complete(_record) do
    # TODO: implement DraftSession.complete
    :ok
  end

  def is_full(_record) do
    # TODO: implement DraftSession.is_full
    {:error, :not_implemented}
  end

  # ── Lifecycle state machine ─────────────────────────────────────────
  @allowed_transitions %{
    "WaitingForPlayers" => ["Drafting", "Abandoned"],
    "Drafting" => ["Completed", "Abandoned"]
  }

  def assert_transition(%__MODULE__{status: current}, to) do
    allowed = Map.get(@allowed_transitions, current, [])
    if to in allowed do
      :ok
    else
      {:error, "Transition #{current} -> #{to} not allowed"}
    end
  end
end
